x-pack/plugins/inference/server/chat_complete/adapters/openai/index.ts

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import OpenAI from 'openai';
+import type {
+  ChatCompletionAssistantMessageParam,
+  ChatCompletionMessageParam,
+  ChatCompletionSystemMessageParam,
+  ChatCompletionToolMessageParam,
+  ChatCompletionUserMessageParam,
+} from 'openai/resources';
+import { filter, from, map, switchMap, tap } from 'rxjs';
+import { Readable } from 'stream';
+import {
+  ChatCompletionChunkEvent,
+  ChatCompletionEventType,
+  Message,
+  MessageRole,
+} from '../../../../common/chat_complete';
+import { createTokenLimitReachedError } from '../../../../common/chat_complete/errors';
+import { createInferenceInternalError } from '../../../../common/errors';
+import { InferenceConnectorAdapter } from '../../types';
+import { eventSourceStreamIntoObservable } from '../event_source_stream_into_observable';
+
+export const openAIAdapter: InferenceConnectorAdapter = {
+  chatComplete: ({ connector, actionsClient, system, messages, toolChoice, tools }) => {
+    const openAIMessages = messagesToOpenAI({ system, messages });
+
+    const toolChoiceForOpenAI =
+      typeof toolChoice === 'string'
+        ? toolChoice
+        : toolChoice
+        ? {
+            function: {
+              name: toolChoice.function,
+            },
+            type: 'function' as const,
+          }
+        : undefined;
+
+    const stream = true;
+
+    const request: Omit<OpenAI.ChatCompletionCreateParams, 'model'> & { model?: string } = {
+      stream,
+      messages: openAIMessages,
+      temperature: 0,
+      tool_choice: toolChoiceForOpenAI,
+      tools: tools
+        ? Object.entries(tools).map(([toolName, { description, schema }]) => {
+            return {
+              type: 'function',
+              function: {
+                name: toolName,
+                description,
+                parameters: (schema ?? {
+                  type: 'object' as const,
+                  properties: {},
+                }) as unknown as Record<string, unknown>,
+              },
+            };
+          })
+        : undefined,
+    };
+
+    return from(
+      actionsClient.execute({
+        actionId: connector.id,
+        params: {
+          subAction: 'stream',
```

Comment:
nit: I know that this will be changed with the Inference Endpoint connector work, so just want to mention here, that it would be better either rename the subAction to some neutral to the type of the response to avoid the confusion (and use stream param in the subAction like it's implemented now https://github.com/elastic/kibana/blob/main/x-pack/plugins/stack_connectors/server/connector_types/openai/openai.ts#L192) or handle "stream" and "non-stream" as separate subActions and move param usage to the level of the subAction implementation.

Reply:
I'm not sure if I fully understand here - at least, I understand the subAction should not be defined by the response type, it should be a parameter (in `subActionParams`). Are there alternatives today? From what I can tell from the sub actions, they either stream or they don't, except `stream`, which has a `stream` parameter which needs to be set to `true` for it to actually stream (which is what I'm doing in the lines below this).

Reply:
Yes, it’s working properly in your implementation. I just prefer a better name for the subAction in the future.